### ISSUE TITLE
Why conformer relative postional encoding left shift have not masked.

### ISSUE THREAD

**Issue by jjjjohnson** [espnet/espnet#2684]
In Transformer-XL postional encoding implementation, we need to mask top right triangle to be zero. while the [code](https://github.com/espnet/espnet/blob/5d97a488c52cd303993064e147fb3a2be20337a5/espnet/nets/pytorch_backend/transformer/attention.py#L200) does not reflect that with default rel_shift(self, x, zero_triu=False). I cannot understand why.

Thanks for any hints!

**Comment by sw005320** [comment]
Thanks for your comments!
This is because it is mainly used for self-attention in the encoder and we do not need to mask the top right triangle for auto-regressive models used in the decoder.
However, the current code is confusing. 
@pengchengguo, could you add some comments to the code?
Also, we may consider the option to handle auto-regressive models in the near future.

**Comment by jjjjohnson** [comment]
Thanks! @sw005320 
I think the whole purpose of rel_shift is to left shift each row of x by (L - row_index), which results in the top right triangle being zero. Based on the Transformer-XL paper it reduces the computation from  O(N2) to O(N). I still can not understand why for self-attention in the encoder, the mask can be omitted...

**Comment by sw005320** [comment]
Sorry, I was talking about different things.
@pengchengguo, could you respond to this?
I also have some difficulties in understanding this relative positional encoding part.
I want to understand more and also want to have some clarifications in the code.

**Comment by pengchengguo** [comment]
No problem, I am checking the paper and code now, and will explain it soon.

**Comment by pengchengguo** [comment]
Hi @jjjjohnson and @sw005320,

Thanks for your suggestion.

The current implementation does exist a bug, but it may not about the top right triangle.

**For the Transformer-xl paper**, they introduce a relative positional encoding `R_{i-j}`, which means the relative distance between a query vector at position `i` and a key vector at position `j`. For an LM related task, `i` should >= `j`, because we can't see future information when training an LM. So the range of `i-j` is from `0` to `L-1` and they force the top right triangle to be zero after `rel_shift`. The resulted term (b) in the paper is like (ignoring the `q^{T}W_{k, R}`):

```
R_{0}, 0, ... , 0
R_{1}, R_{0}, ... , 0
...
R_{L-1}, R_{L-2}, ... , R_{0}
```

**For ASR**, we can capture future information, which means the range of `i-j` is from `-(L-1)` to `L-1`. In the current implementation, we want to use the absolute value of `i-j` and make the same range of original paper. The generated term (b) becomes like

```
R_{0}, R_{1}, ... , R_{L-1}
R_{1}, R_{0}, ... , R_{L-2}
...
R_{L-1}, R_{L-2}, ... , R_{0}
```

The original `rel_shift` method is not right for the above computation. we will fix the bug ASAP and just try to use range from `-(L-1)` to `L-1`.

BTW, many thanks if you cloud give me some suggestions about capturing both left and right information with relative positional encoding.

**Comment by jjjjohnson** [comment]
Thank @pengchengguo for the clear explanation!
For shift matrix_bd, you can build a index array and gather it like this:
 
![image](https://user-images.githubusercontent.com/20569095/99337487-52eb1480-28bd-11eb-8e21-e2e7f7404505.png)

**Comment by pengchengguo** [comment]
Great! Thanks so much. We will try it.

**Comment by placebokkk** [comment]
Another way is to construct a (2L -1) length position embedding R for [-L+1, L-1]. But This may increase computation cost.

[
q_{0}*R_{-L+1}, q_{0}*R_{-L+2}, ... , q_{0}*R_{0}, ..., q_{0}*R_{L-2}, q_{0}*R_{L-1},
q_{1}*R_{-L+1}, q_{1}*R_{-L+2}, ... , q_{1}*R_{0}, ..., q_{1}*R_{L-2}, q_{1}*R_{L-1},
...
q_{L-1}*R_{-L+1}, q_{L-1}*R_{-L+2}, ... , q_{L-1}*R_{0}, ..., q_{L-1}*R_{L-2}, q_{L-1}*R_{L-1},
]

Applying current rel_shift and will return

[
q_{0}*R_{0}, ..., q_{0}*R_{L-2}, q_{0}*R_{L-1}, 0 , somethings
q_{1}*R_{-1}, q_{1}*R_{0}, ..., q_{1}*R_{L-2}, somethings
...
q_{L-1}*R_{-L+1}, q_{L-1}*R_{-L+2},  q_{L-1}*R_{0}, somethings
]

Then take the first L columns.

**Comment by pengchengguo** [comment]
Hi @placebokkk,

Thanks so much for your suggestions. Actually, we have re-implemented the relative positional encoding the same way as your idea, and this paper https://arxiv.org/pdf/2005.09940.pdf. Now, we are conducting some experiments to verify the implementation before making the PR.

**Comment by placebokkk** [comment]
> Hi @placebokkk,
> 
> Thanks so much for your suggestions. Actually, we have re-implemented the relative positional encoding the same way as your idea, and this paper https://arxiv.org/pdf/2005.09940.pdf. Now, we are conducting some experiments to verify the implementation before making the PR.

Great! Thanks for this information.

**Comment by jaesong** [comment]
FYI, there's another way to construct the relative position matrix using `torch.as_strided()`:

```python
import torch


torch.manual_seed(1)
batch, head, max_time, dim = 1, 1, 3, 1

# x[:, :, t + max_time]: relative embedding of t (-max_time <= t <= max_time)
x = torch.randn(batch, head, max_time * 2 + 1, dim)
x_stride = tuple(x.stride())
x_size = tuple(x.size())
print('x:', x)

# split the time axis into two axes with same stride
axis = 2
new_stride = x_stride[:axis] + (x_stride[axis], x_stride[axis]) + x_stride[3:]
new_size = x_size[:axis] + (x_size[axis] // 2 + 1, x_size[axis] // 2 + 1) + x_size[3:]
# xx.shape == (1, 1, 4, 4, 1)
# xx[:, :, i, j] == x[:, :, (i + j)]
xx = torch.as_strided(x, new_size, new_stride)
print('xx:', xx)

# x_relative[:, :, i, j] == xx[:, :, -i, j] == x[:, :, (j - i + max_time)]
# this gives the relative embedding matrix
x_relative = torch.flip(xx, [axis])

assert x_relative[:, :, 0, 0] == x[:, :, max_time]
assert x_relative[:, :, 1, 1] == x[:, :, max_time]

assert x_relative[:, :, 0, 1] == x[:, :, max_time + 1]
assert x_relative[:, :, 1, 2] == x[:, :, max_time + 1]

assert x_relative[:, :, 2, 0] == x[:, :, max_time - 2]
assert x_relative[:, :, 3, 1] == x[:, :, max_time - 2]

# xx is reshape of x, thus no extra memory is allocated
assert x.storage().data_ptr() == xx.storage().data_ptr()
# 7
print('x allocated memory:', x.storage().size())
# 7 (but note that `xx.numel() == 16`)
print('xx allocated memory:', xx.storage().size())
# flip() allocates new memory. this would be an issue if `max_time` is really big.
# 16
print('x_relative allocated memory:', x_relative.storage().size())
```

Dynamic convolution uses similar trick. I'm not aware of any implementation using this trick for relative attention though.

**Comment by jaesong** [comment]
Anyway, the current implementation seems to work as intended, so I guess the original issue is fixed?
